Patch errors for Asahi Linux.

zaneenders opened the pull request:
Hi comnerd, 

I was checking out [VirtualTerminal](https://github.com/compnerd/VirtualTerminal) and ran into a few errors. I'm not 100% sure if these fixes are correct as i'm still new to Linux, but hopefully it's helpful.

Error I got for POSIXError.
```console
└╴  Value of optional type 'UnsafeMutablePointer<CChar>?' (aka 'Optional<UnsafeMutablePointer<Int8>>') must be unwrapped to a value of type 'UnsafeMutablePointer<CChar>' (aka 'UnsafeMutablePointer<Int8>') SourceKit  [23, 33]
```

Error I got for SystemInfo

```console
└╴  Cannot convert return expression of type '()' to return type 'Int' SourceKit  [15, 5]
```

My system info.

```console 
➜  swift-platform-core git:(zane-asahi-linux-patch) hostnamectl
 Static hostname: fedora
       Icon name: computer
      Machine ID: 29196fa12af0453e852f15fe4c684c25
         Boot ID: ea543588fb8240fb95363a0e1e2c8ae9
Operating System: Fedora Linux Asahi Remix 42 (Forty Two [Adams])
     CPE OS Name: cpe:/o:fedoraproject:fedora-asahi-remix:42
          Kernel: Linux 6.14.8-400.asahi.fc42.aarch64+16k
    Architecture: arm64
    
➜  swift-platform-core git:(zane-asahi-linux-patch) swift --version
Swift version 6.1.2 (swift-6.1.2-RELEASE)
Target: aarch64-unknown-linux-gnu
```

Thanks,
Zane

compnerd committed:
WindowsCore: add a couple more type cast constants

Introduce `CT_CTYPE3` and `C3_FULLWIDTH` wrappers.

compnerd commented on Sources/POSIXCore/POSIXError.swift:
```diff
@@ -3,30 +3,34 @@
+  extension POSIXError: CustomStringConvertible {
+    public var description: String {
+      return withUnsafeTemporaryAllocation(of: CChar.self, capacity: 256) {
+        guard let _baseAddres = $0.baseAddress else {
```
I think that this will break on macOS.

compnerd replied:
Ugh, never mind, we already unwrap afterwards. Okay, hoisting the `baseAddress` makes sense.

compnerd reviewed (commented):
Thanks for the pull request! Can we avoid the whitespace changes please? Let me check the changes on macOS as we might need to make this specific to Linux.

compnerd commented on Sources/POSIXCore/POSIXError.swift:
```diff
@@ -3,30 +3,34 @@
+  extension POSIXError: CustomStringConvertible {
+    public var description: String {
+      return withUnsafeTemporaryAllocation(of: CChar.self, capacity: 256) {
+        guard let _baseAddres = $0.baseAddress else {
+          return "POSIX Error missing baseAddress"
+        }
+        guard strerror_r(code, _baseAddres, $0.count) == 0,
+          let baseAddress = $0.baseAddress
+        else {
+          return "POSIX Error \(code)"
+        }
```
```suggestion
        guard let baseAddres = $0.baseAddress,
            strerror_r(code, _baseAddres, $0.count) == 0 else {
          return "POSIX Error \(code)"
        }
}
```

compnerd replied:
This is more succinct and corrects the typo.

zaneenders replied:
Thank you for showing me this.

compnerd commented:
https://github.com/compnerd/swift-platform-core/pull/3 would be the approach that I would recommend for this. If you stack this on top of that, you should be able to simply remove the `#if os(macOS)` and the `#else ...` case for the page size.

zaneenders committed:
Fix Glic imports.

zaneenders committed:
Fix macOS build.

zaneenders commented:
@compnerd thanks for showing me `traits` and package variables. I haven't used those before.

I rebased my branch using your overlay (also a new term for me). 

When checking my branch on the macOS side I got the following errors in SignalHandler.swift. Which I fixed by changing line 27.

1. New Swift parser generated errors for code that C++ parser accepted
```
// Copyright © 2025 Saleem Abdulrasool <compnerd@compnerd.org>
```
Kinda of an odd parser error.


1. Expected '>' to end generic argument clause
2. Unexpected code '2, CInt>(repeating: -1)' before conditional compilation clause
```
#if $InlineArray
  private nonisolated(unsafe) static var fds = InlineArray<2, CInt>(repeating: -1)
#else
  private nonisolated(unsafe) static var fds = Array<CInt>(repeating: -1, count: 2)
#endif
```

```
swift --version
Apple Swift version 6.1.2 (swift-6.1.2-RELEASE)
Target: arm64-apple-macosx15.0
```

zaneenders commented:
Also planning on squashing commits.

zaneenders commented on Sources/POSIXCore/SignalHandler.swift:
```diff
@@ -3,6 +3,10 @@
 
+#if canImport(Glibc)
+import Glibc
+#endif
```
Ok, so `sigaction`, `errno`, `close`, `pipe`,`fcntl`, ` F_GETFL`, ` O_NONBLOCK`, `read`, `EINTR`, `close`, `write`, ` sigemptyset`, ` SA_RESTART`, all throw compiler errors for not being available. 

Sigaction i'm the most confused how to overlay as these are the two types Glibc is using. Do you have an example or ideas on what do about the struct?

```swift 
package typealias sigaction = Glibc.sigaction // didn't work
```

```swift
// I think a pattern like this would work for all the functions.
package func sigaction(_ __sig: Int32, _ __act: UnsafePointer<sigaction>!, _ __oact: UnsafeMutablePointer<sigaction>!) -> Int32 {
    Glibc.sigaction(__sig, __act, __oact)
}
```

```swift 
// But really not sure what do do here.
public struct sigaction {

    public init()

    public init(__sigaction_handler: sigaction.__Unnamed_union___sigaction_handler, sa_mask: __sigset_t, sa_flags: Int32, sa_restorer: (@convention(c) () -> Void)!)

    public struct __Unnamed_union___sigaction_handler {

        public init(sa_handler: __sighandler_t!)

        public init(sa_sigaction: (@convention(c) (Int32, UnsafeMutablePointer<siginfo_t>?, UnsafeMutableRawPointer?) -> Void)!)

        public init()

        public var sa_handler: __sighandler_t!

        public var sa_sigaction: (@convention(c) (Int32, UnsafeMutablePointer<siginfo_t>?, UnsafeMutableRawPointer?) -> Void)!
    }

    public var __sigaction_handler: sigaction.__Unnamed_union___sigaction_handler

    public var sa_mask: __sigset_t

    public var sa_flags: Int32

    public var sa_restorer: (@convention(c) () -> Void)!
}
```

compnerd commented:
I've applied a simpler variation on all of these points. Please give it another whirl and if you find that there are still issues, please file an issue/PR.

BTW, I've tested the current version using `swift build --traits GNU` and verified it builds.
